### PR TITLE
1270 window refactor

### LINKED ISSUE #1270: screen thermal effect
Most screens have an air layer behind them that heats up and heat is able to transfer through the window into the zone. This is not modelled and causes an underestimation of the indirect solar heat gains.

### PR DESCRIPTION
**Pull request by Mathadon**
closes #1270

### CODE REVIEW

**Comment by Mathadon** [comment]
@ettorez I have refactored the shading model such that the window model now looks like this:

image

Have a look at `IDEAS.Buildings.Components.Shading.Interfaces.PartialShadingDevice`. You should be able to duplicate this model and revise the heat exchange blocks to integrate your model equations into a shading block. 

I still have to integrate the fluid port into the shading.

**Comment by EttoreZ** [comment]
> @EttoreZ I have refactored the shading model such that the window model now looks like this:
> 
> image
> 
> Have a look at `IDEAS.Buildings.Components.Shading.Interfaces.PartialShadingDevice`. You should be able to duplicate this model and revise the heat exchange blocks to integrate your model equations into a shading block.
> 
> I still have to integrate the fluid port into the shading.

Thank you Filip @Mathadon! I will take a look.

File: IDEAS/Buildings/Components/Shading/Interfaces/PartialShadingDevice.mo [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+protected
+  Modelica.Blocks.Interfaces.RealInput TDryBul_internal = Te_internal
+    "Internal variable to avoid assigning a value to a non-input in the model extends statement";
+  Modelica.Blocks.Interfaces.RealInput TEnv_internal
+    "Internal variable to avoid assigning a value to a non-input in the model extends statement";
+equation
+  connect(TEnv_internal,TEnv);
+  if not haveBoundaryPorts then
+    TEnv_internal = 273.15;
+  end if;
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Interfaces/PartialShadingDevice.mo:49]: Shouldn't the connection be part of the if-statement (`if not haveboundaryPorts then` constant `else` connect)? Or is this a more clean implementation?

File: IDEAS/Buildings/Components/Shading/Interfaces/ShadingProperties.mo [hunk 2]
```diff
@@ -2,7 +2,10 @@ within IDEAS.Buildings.Components.Shading.Interfaces;
 record ShadingProperties
   "Default: no shading"
   extends Modelica.Icons.Record;
-  parameter Boolean controlled = false
+  parameter Boolean controlled = shaType==
+      IDEAS.Buildings.Components.Shading.Interfaces.ShadingType.Screen or shaType==
+      IDEAS.Buildings.Components.Shading.Interfaces.ShadingType.BoxAndScreen or shaType==
+      IDEAS.Buildings.Components.Shading.Interfaces.ShadingType.OverhangAndScreen
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Interfaces/ShadingProperties.mo:8]: More the ` or shaType==` at the end of every line to the start of the next line to improve readability.
E.g.
`controlled = `
`shaType == IDEAS.Buildings.Components.Shading.Interfaces.ShadingType.Screen`
`or shaType == IDEAS.Buildings.Components.Shading.Interfaces.ShadingType.BoxAndScreen`

File: IDEAS/Buildings/Components/Shading/OverhangAndScreen.mo [hunk 3]
```diff
@@ -44,6 +44,10 @@ initial equation
 </html>", revisions="<html>
 <ul>
 <li>
+July 18, 2022 by Filip Jorissen:<br/>
+Refactored for #1270 for including thermal effect of screens.
+</li>
+<li>
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/OverhangAndScreen.mo:50]: Only change in this file is the revision history. Did I miss a change, or should this revision be added to another model?

File: IDEAS/Buildings/Components/Shading/Screen.mo [hunk 4]
```diff
@@ -1,13 +1,26 @@
...
 model Screen "Controllable exterior screen"
-  extends IDEAS.Buildings.Components.Shading.Interfaces.PartialShading(
-                                                             final controlled=true);
+  extends IDEAS.Buildings.Components.Shading.Interfaces.PartialShadingDevice(
+    TSha = TShaScreen,
+    TDryBul_internal = limiter.y*TSha + (1-limiter.y)*Te_internal,
+    epsSw_shading = 1 - shaCorr,
+    final controlled = true,
+    TEnvExpr(y = TEnv_screen),
+    TeExpr(y = TDryBul_internal));
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Screen.mo:9]: Are the new parameter declarations of the extension deliberately not set to final?

File: IDEAS/Buildings/Components/Shading/Shading.mo [hunk 5]
```diff
@@ -9,7 +9,21 @@ model Shading
...
+    epsLw_frame=epsLw_frame,
+    epsLw_glazing=epsLw_glazing,
+    epsSw_shading=epsSw_shading,
+    final g_glazing=g_glazing,
+    linCon=linCon,
+    linRad=linRad,
     azi=azi,
+    haveBoundaryPorts=true,
+    haveFrame=haveFrame,
+    final hSha=hSha,
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Shading.mo:26]: Some parameters are not final, why? (Similar comment/questions for the other protected components.)

File: IDEAS/Buildings/Components/Shading/Shading.mo [hunk 6]
```diff
@@ -53,27 +109,83 @@ protected
...
+    A_frame=A_frame,
+    inc=inc,
+    Tenv_nom=Tenv_nom,
+    epsSw_frame=epsSw_frame,
+    epsLw_frame=epsLw_frame,
+    epsLw_glazing=epsLw_glazing,
+    epsSw_shading=epsSw_shading,
+    final g_glazing=g_glazing,
+    linCon=linCon,
+    linRad=linRad,                             azi=azi,
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Shading.mo:165]: Put `azi=azi,` at a new line.

File: IDEAS/Buildings/Components/Window.mo [hunk 7]
```diff
@@ -54,21 +54,32 @@ model Window "Multipane window"
...
+      A_frame = A * frac,
+      A_glazing = A * (1 - frac),
+      Tenv_nom = sim.Tenv_nom,
+      epsLw_frame = fraType.mat.epsLw,
+      epsLw_glazing = layMul.parEpsLw_b,
+      epsSw_frame = fraType.mat.epsSw,
+      g_glazing=glazing.g_value,
+      inc = incInt,
+      linCon = linExtCon or sim.linearise,
+      linRad = linExtRad or sim.linearise,
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Window.mo:69]: Deliberately not set to final?

File: IDEAS/Buildings/Components/Shading/Interfaces/PartialShading.mo [hunk 8]
```diff
@@ -3,11 +3,38 @@ partial model PartialShading "Window shading partial"
...
+  parameter Modelica.Units.SI.Angle inc "Surface inclination";
+  parameter Modelica.Units.SI.Temperature Tenv_nom=280
+    "Nominal temperature of environment";
+  parameter Modelica.Units.SI.Emissivity epsSw_frame
+    "Short wave solar absorption coefficient of the frame";
+  parameter Modelica.Units.SI.Emissivity epsLw_frame
+    "Long wave solar absorption coefficient of the frame";
+  parameter Modelica.Units.SI.Emissivity epsLw_glazing
+    "Long wave solar absorption coefficient of the glazing";
+  parameter Modelica.Units.SI.Emissivity epsSw_shading = 0.5
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Interfaces/PartialShading.mo:17]: Why is this the only emissivity parameter for which a default value is provided?

File: IDEAS/Buildings/Components/Shading/Screen.mo [hunk 9]
```diff
@@ -1,13 +1,26 @@
...
 
-  parameter Real shaCorr=0.24 "Shortwave transmittance of shortwave radiation";
+  parameter Real shaCorr(min=0, max=1) = 0.24
+    "Shortwave transmittance of the screen";
 
 protected
-  Modelica.Blocks.Nonlinear.Limiter limiter(uMax=1, uMin=0);
-
+  Modelica.Units.SI.Temperature TEnv_screen = limiter.y*TSha + (1-limiter.y)*TEnv_internal;
```
Comment (jelgerjansen) [inline IDEAS/Buildings/Components/Shading/Screen.mo:15]: A documentation string should be provided.

File: IDEAS/Buildings/Components/BaseClasses/ConductiveHeatTransfer/MultiLayer.mo [hunk 10]
```diff
@@ -32,6 +32,11 @@ model MultiLayer "multiple material layers in series"
     annotation(Dialog(enable=linIntCon));
   parameter Boolean checkCoatings = false
     "Check whether air layers have a coating";
+  final parameter Modelica.Units.SI.Emissivity parEpsSw_a = mats[nLay].epsSw_a;
+  final parameter Modelica.Units.SI.Emissivity parEpsSw_b =  mats[1].epsSw_b;
+  final parameter Modelica.Units.SI.Emissivity parEpsLw_a =  mats[nLay].epsLw_a;
+  final parameter Modelica.Units.SI.Emissivity parEpsLw_b =  mats[1].epsLw_b;
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/BaseClasses/ConductiveHeatTransfer/MultiLayer.mo:38]: Yup, good catch!

File: IDEAS/Buildings/Components/BaseClasses/ConvectiveHeatTransfer/ExteriorConvection.mo [hunk 11]
```diff
@@ -12,14 +12,14 @@ model ExteriorConvection "exterior surface convection"
 
   Modelica.Thermal.HeatTransfer.Interfaces.HeatPort_a port_a
     "Connector for solid part"
-    annotation (Placement(transformation(extent={{-110,-10},{-90,10}})));
+    annotation (Placement(visible = true, transformation(extent = {{-110, -10}, {-90, 10}}, rotation = 0), iconTransformation(extent = {{-110, -10}, {-90, 10}}, rotation = 0)));
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/BaseClasses/ConvectiveHeatTransfer/ExteriorConvection.mo:15]: That's part of changes due to OpenModelica. It's a hassle to remove it manually so I guess that we're stuck with it in the long run. Fortunately, OM does not rewrite the whole file. Chances are that Dymola will undo these changes but let's use OM for development then.

File: IDEAS/Buildings/Components/BaseClasses/ConvectiveHeatTransfer/ExteriorConvection.mo [hunk 12]
```diff
@@ -84,12 +84,7 @@ equation
   port_a.Q_flow = A * hConExt * dT;
 
   annotation (
-    Icon(graphics={
-        Rectangle(
-          extent={{-90,80},{-60,-80}},
-          fillColor={192,192,192},
-          fillPattern=FillPattern.Backward,
-          pattern=LinePattern.None),
+    Icon(graphics={Rectangle(fillColor = {192, 192, 192}, pattern = LinePattern.None, fillPattern = FillPattern.Backward, extent = {{-90, 80}, {-60, -80}}),
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/BaseClasses/ConvectiveHeatTransfer/ExteriorConvection.mo:87]: Indeed, result of OM.

File: IDEAS/Buildings/Components/Examples/BuildingShadeExample.mo [hunk 13]
```diff
@@ -13,10 +21,26 @@ protected
...
 public
-  Shading.None none(azi=azi.k)
+  Shading.None none(
+    A_glazing=0,
+    A_frame=0,
+    inc=0,
+    epsSw_frame=1,
+    epsLw_frame=1,
+    epsLw_glazing=1,
+    g_glazing=0,    azi=azi.k,
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/Examples/BuildingShadeExample.mo:31]: again, a result of OpenModelica formatting.. but I'll fix it

[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/Examples/BuildingShadeExample.mo:31]:
```suggestion
    g_glazing=0,    
    azi=azi.k,
```

File: IDEAS/Buildings/Components/Examples/BuildingShadeExample.mo [hunk 14]
```diff
@@ -29,22 +53,46 @@ public
...
     annotation (Placement(transformation(extent={{-100,20},{-80,40}})));
-  Shading.Screen screen(azi=azi.k)
+  Shading.Screen screen(
+    A_glazing=0,
+    A_frame=0,
+    inc=0,
+    epsSw_frame=1,
+    epsLw_frame=1,
+    epsLw_glazing=1,
+    g_glazing=0,        azi=azi.k,
```
[the remark this answers is not in the document]
Comment (Mathadon) [reply IDEAS/Buildings/Components/Examples/BuildingShadeExample.mo:64]:
```suggestion
    g_glazing=0,
    azi=azi.k,
```

File: IDEAS/Buildings/Components/Shading/Shading.mo (hunk 5 above)
Comment (Mathadon) [reply IDEAS/Buildings/Components/Shading/Shading.mo:26]: For no particular reason tbh. I removed the final for consistency.

File: IDEAS/Buildings/Components/Shading/Shading.mo (hunk 6 above)
Comment (Mathadon) [reply IDEAS/Buildings/Components/Shading/Shading.mo:165]: done

File: IDEAS/Buildings/Components/Shading/Interfaces/PartialShading.mo (hunk 8 above)
Comment (Mathadon) [reply IDEAS/Buildings/Components/Shading/Interfaces/PartialShading.mo:17]: I probably had to do it while testing and then forgot to remove it, thanks!

File: IDEAS/Buildings/Components/Shading/Screen.mo (hunk 9 above)
Comment (Mathadon) [reply IDEAS/Buildings/Components/Shading/Screen.mo:15]: yup

**Pull request merged**